From c4d58b6ff1ebee003131dd02033a485493070c58 Mon Sep 17 00:00:00 2001
From: "Lu.Jiang" <lu.jiang@windriver.com>
Date: Thu, 4 Sep 2014 18:29:58 +0800
Subject: [PATCH 0177/1089] eSDHC: Add clock glitch impact list for errata
 A-003980

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Add impact list for clock glitch errata.
A-003980: SDHC: Clock glitch is generated on the card clock when software
reset or clock divider change was implemented, all command will be timeout
in this situation. This issue can be occured when card setup and software
reset all command be issued.
As errata said, this issue only exist on below board:

T4240-R1.0 B4860-R1.0 P3041-R1.0
P3041-R2.0 P2041-R1.0 P2041-R1.1
P2041-R2.0 P1010-R1.0

Signed-off-by: Jerry Huang <Chang-Ming.Huang@freescale.com>
Signed-off-by: Haijun Zhang <Haijun.Zhang@freescale.com>
Change-Id: Ia4461974f7f08fe57997ca77036a0eeda57bea74
Reviewed-on: http://git.am.freescale.net:8181/918
Reviewed-by: Xie Xiaobo-R63061 <X.Xie@freescale.com>
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Signed-off-by: Lu.Jiang <lu.jiang@windriver.com>
---
 drivers/mmc/host/sdhci-of-esdhc.c | 17 +++++++++++++++++
 drivers/mmc/host/sdhci-pltfm.c    |  7 -------
 2 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/drivers/mmc/host/sdhci-of-esdhc.c b/drivers/mmc/host/sdhci-of-esdhc.c
index 13f25e1..12d7a44 100644
--- a/drivers/mmc/host/sdhci-of-esdhc.c
+++ b/drivers/mmc/host/sdhci-of-esdhc.c
@@ -388,6 +388,23 @@ static void esdhc_of_platform_init(struct sdhci_host *host)
 
 	if (vvn > VENDOR_V_22)
 		host->quirks &= ~SDHCI_QUIRK_NO_BUSY_IRQ;
+
+	/*
+	 * Check for A-005055: A glitch is generated on the card clock
+	 * due to software reset or a clock change
+	 * Impact list:
+	 * T4240-R1.0 B4860-R1.0 P3041-R1.0 P3041-R2.0 P2041-R1.0
+	 * P2041-R1.1 P2041-R2.0 P1010-R1.0
+	 */
+	if ((fsl_svr_is(SVR_T4240) && fsl_svr_rev_is(1, 0)) ||
+		(fsl_svr_is(SVR_B4860) && fsl_svr_rev_is(1, 0)) ||
+		(fsl_svr_is(SVR_P3041) && fsl_svr_rev_is(1, 0)) ||
+		(fsl_svr_is(SVR_P3041) && fsl_svr_rev_is(2, 0)) ||
+		(fsl_svr_is(SVR_P2041) && fsl_svr_rev_is(2, 0)) ||
+		(fsl_svr_is(SVR_P2041) && fsl_svr_rev_is(1, 1)) ||
+		(fsl_svr_is(SVR_P2041) && fsl_svr_rev_is(1, 0)) ||
+		(fsl_svr_is(SVR_P1010) && fsl_svr_rev_is(1, 0)))
+		host->quirks2 |= SDHCI_QUIRK2_BROKEN_RESET_ALL;
 }
 
 static int esdhc_pltfm_bus_width(struct sdhci_host *host, int width)
diff --git a/drivers/mmc/host/sdhci-pltfm.c b/drivers/mmc/host/sdhci-pltfm.c
index 4553355..8412d30 100644
--- a/drivers/mmc/host/sdhci-pltfm.c
+++ b/drivers/mmc/host/sdhci-pltfm.c
@@ -99,19 +99,12 @@ void sdhci_get_of_property(struct platform_device *pdev)
 			host->quirks |= SDHCI_QUIRK_BROKEN_TIMEOUT_VAL;
 
 		if (of_device_is_compatible(np, "fsl,t4240-esdhc")) {
-			host->quirks2 |= SDHCI_QUIRK2_BROKEN_RESET_ALL;
 			host->quirks2 |= SDHCI_QUIRK2_LONG_TIME_CMD_COMPLETE_IRQ;
 			host->quirks2 |= SDHCI_QUIRK2_CIRCUIT_SUPPORT_VS33;
 			host->quirks2 |= SDHCI_QUIRK2_FORCE_CMD13_DETECT_CARD;
 			host->quirks |= SDHCI_QUIRK_BROKEN_ADMA;
 		}
 
-		if (of_device_is_compatible(np, "fsl,p4860-rev1-esdhc") ||
-		    of_device_is_compatible(np, "fsl,p1010-esdhc") ||
-		    of_device_is_compatible(np, "fsl,p2041-esdhc") ||
-		    of_device_is_compatible(np, "fsl,p3041-esdhc"))
-			host->quirks2 |= SDHCI_QUIRK2_BROKEN_RESET_ALL;
-
 		clk = of_get_property(np, "clock-frequency", &size);
 		if (clk && size == sizeof(*clk) && *clk)
 			pltfm_host->clock = be32_to_cpup(clk);
-- 
2.0.2

